Add rollback auto update

afragen opened the pull request:
This PR adds the rollback auto update functionality to core. This is already part of the [Rollback Update Failure](https://wordpress.org/plugins/rollback-update-failure) feature plugin.

This requires [PR2225](https://github.com/WordPress/wordpress-develop/pull/2225). ✅

Trac ticket: https://core.trac.wordpress.org/ticket/58281

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

afragen committed:
add rollback auto update

afragen committed:
update from plugin

afragen committed:
Merge branch 'trunk' into rollack-auto-update

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+	 *
+	 * @var bool
+	 */
+	private $no_error = false;
```
```suggestion
	private $update_is_safe = false;
```
It might be worth being a bit more explicit here to avoid confusion when reading implementation and thinking "no" means the number of errors, etc.

> Stores whether the update is safe to proceed without Rollback.

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+
+		if ( is_wp_error( $response ) ) {
+			// If it isn't possible to run the check, assume an error.
+			throw new \Exception( $response->get_error_message() );
```
```suggestion
			throw new Exception( $response->get_error_message() );
```
Leftover from the feature plugin.

:1234: Applies elsewhere in the PR (in docs too).

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+
+		error_log( $hook_extra['plugin'] . ' processing...' );
+
+		/**
```
```suggestion
		/*
```

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+			return $result;
+		}
+
+		error_log( $hook_extra['plugin'] . ' processing...' );
```
Definitely worth having these logs for testing. Just a reminder for us to remove these before merge.

afragen replied:
Remove of put behind a WP_DEBUG is true?

costdev replied:
Don't think it needs to be shown with or without `WP_DEBUG` after merge, but for testing it might as well just get logged every time someone tests the PR

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+
+	/**
+	 * Stores handler parameters.
+	 *
```
Let's add some `@since` on these properties and the methods

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+	 * If an error is found, the previously installed version will be reinstalled
+	 * and an email will be sent to the site administrator.
+	 *
+	 * @param string $plugin The plugin to check.
```
Needs a `@global` for `$wp_filesystem`

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+		$this->no_error = 200 === $code;
+
+		if ( str_contains( $body, 'wp-die-message' ) || 200 !== $code ) {
+			$error = new WP_Error(
+				'new_version_error',
+				sprintf(
+					/* translators: %s: The name of the plugin. */
+					__( 'The new version of %s contains an error' ),
+					get_plugin_data( WP_PLUGIN_DIR . '/' . $plugin )['Name']
+				)
+			);
+			throw new \Exception( $error->get_error_message() );
```
Should this just send the message directly to `new Exception()` instead?

afragen replied:
something tells me `throw new Exception( new WP_Error()...` might be a little less clear.

costdev replied:
I meant more:

```php
throw new Exception(
	sprintf(
		/* translators: %s: The name of the plugin. */
		__( 'The new version of %s contains an error' ),
		get_plugin_data( WP_PLUGIN_DIR . '/' . $plugin )['Name']
	)
);
```

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+
+		$this->cron_rollback();
+
+		/**
```
```suggestion
		/*
```

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+	/**
+	 * Skip (exit handler() early) for non-fatal errors or non-errors.
+	 *
+	 * @return bool
```
The return needs a description

afragen replied:
Not sure what you mean. Isn't the first line a description?

costdev replied:
We usually have a description for the `@return` for Devhub.

Like:
`@return bool Whether to skip for non-fatal errors or non-errors.`

![image](https://user-images.githubusercontent.com/79332690/224852919-bfc3f026-a270-4405-aa2e-fe3628a58c2b.png)

costdev reviewed (changes requested):
Thanks for prepping the PR @afragen! I've left a few thoughts in this review

afragen committed:
Update src/wp-admin/includes/class-rollback-auto-update.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+	 * @return bool
+	 */
+	private function non_fatal_errors() {
+		$e                = error_get_last();
+		$non_fatal_errors = ( ! empty( $e ) && $this->error_types !== $e['type'] );
```
```suggestion
		$last_error       = error_get_last();
		$non_fatal_errors = ( ! empty( $last_error ) && $this->error_types !== $last_error['type'] );
```
We've been generally trying to avoid single-character variable names for the sake of readability.

costdev replied:
Also, since `$non_fatal_errors` is `bool`, maybe `$has_non_fatal_errors` is clearer. I thought for a second this held the non-fatal errors.

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+		}
+
+		$automatic_upgrader      = new WP_Automatic_Updater();
+		$send_plugin_theme_email = new \ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
+		$send_plugin_theme_email->setAccessible( true );
+		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
```
```suggestion
		$send_plugin_theme_email = new \ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
		$send_plugin_theme_email->setAccessible( true );
		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
		$send_plugin_theme_email->setAccessible( false );
```
Reset the accessibility after we're done with this Core class, or we may not need this reflection for merge, not sure. Are we planning to duplicate `WP_Automatic_Updater::send_plugin_theme_email()`, or abstract it someway so it's a shared method? 🤔 Not sure of the best path.

afragen replied:
I don't understand?

costdev replied:
To avoid using the Reflection API for this bit, I think we might be better duplicating `WP_Automatic_Updater::send_plugin_theme_email()` into this class for merge, and we can explore ways to reduce code duplication later.

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+	 *
+	 * @return array
+	 */
+	public function props( $email, $type, $successful_updates, $failed_updates ) {
```
Method might need an update/removal for a Core merge

afragen replied:
Why?

costdev replied:
Sorry, I should have elaborated. Since this method adds the failure message, it's not really "props", or "credit where credit is due". It's more "Adds a message to note that failures occurred and rollbacks may have been performed.", and `public function add_rollback_message_to_email()`, for example (NTIH)

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+	private function check_plugin_for_errors( $plugin ) {
+		global $wp_filesystem;
+
+		if ( file_exists( ABSPATH . '.maintenance' ) ) {
```
```suggestion
		if ( $wp_filesystem->exists( ABSPATH . '.maintenance' ) ) {
```
We're using `$wp_filesystem` anyway

afragen committed:
update docblocks

afragen committed:
no escaping needed

afragen committed:
update variable name for clarity

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,405 @@
+		$automatic_upgrader      = new WP_Automatic_Updater();
+		$send_plugin_theme_email = new \ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
+		$send_plugin_theme_email->setAccessible( true );
+		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
```
After this, should we remove the filter added at the start of this method?
```php
remove_filter( 'auto_plugin_theme_update_email', array( $this, 'props' ), 10, 4 );
```

afragen replied:
Shouldn't we remove it at the beginning of `props()`?

costdev replied:
Could do, although that assumes we'll only ever use `::props()` in this filter. I mean, I imagine that will be the case, but might be worth just removing it in the same method we add it. No strong feelings one way or the other.

afragen committed:
Merge remote-tracking branch 'afragen/rollack-auto-update' into rollack-auto-update

afragen committed:
Update src/wp-admin/includes/class-rollback-auto-update.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
Update src/wp-admin/includes/class-rollback-auto-update.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
update since

afragen committed:
description for @return

afragen committed:
remove filter at end of function calling it

afragen committed:
more updates

afragen committed:
rename -- again

afragen committed:
remove error logging for commit

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,440 @@
+		}
+
+		// Already processed.
+		if ( in_array( $hook_extra['plugin'], self::$processed, true ) ) {
+			return $result;
+		}
```
Since a lookup is faster than `in_array()`, a slight performance improvement for bulk auto-updates could be:
```php
if ( isset( self::$processed[ $hook_extra['plugin'] ] ) ) {
```

Then instead of:
```php
self::$processed[] = $hook_extra['plugin'];
```
use:
```php
self::$processed[ $hook_extra['plugin'] ] = true;
```

afragen committed:
Merge branch 'trunk' into rollack-auto-update

afragen committed:
pass WP_Upgrader as param

allows for cleanup of filter hooks from fatal plugin updates

afragen committed:
restart theme auto-updates after plugin fatal update

afragen committed:
add @since

afragen committed:
remove line

afragen committed:
cleanup and get update transients if need to restart.

afragen committed:
Merge branch 'trunk' into rollack-auto-update

afragen committed:
moving the deck chairs

kebbet commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,493 @@
+			return $email;
+		}
+		$body   = explode( "\n", $email['body'] );
+		$failed = __( 'These plugins failed to update or may have been rolled back due to detection of a fatal error:' );
```
Is ”rolled back” a understandable term to show a user? It is really hard to translate in a meaningfull way. Can the sentence be:
`These plugins failed to update or may have been restored due to detection of a fatal error:` ?

kebbet replied:
Or even:
`These plugins failed to update or may have been restored to their previous version due to detection of a fatal error:`

costdev replied:
While the feature and the feature project are called **Rollback**, I agree that the terminology is not very accessible to most non-technical users.


I'd suggest that we continue discussion on [this ticket](https://core.trac.wordpress.org/ticket/58282) so that discussion and decision-making happen in one place, then we can apply the decision(s) to this PR and related **Rollback** user-facing strings.

kebbet replied:
This string was the thing that got me going, so yeah, await a decision there and maybe apply changes here post decision.

NekoJonez replied:
"Returned" or "not changed"... That's what I'd use.

afragen committed:
improved language

afragen committed:
Merge branch 'trunk' into rollack-auto-update

afragen committed:
update for parity with plugin

afragen committed:
Merge branch 'trunk' into rollack-auto-update

afragen committed:
refactor email data gathering a bit

afragen committed:
not for PR

dream-encode commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,499 @@
+ *
+ * @package WordPress
+ * @subpackage Administration
+ * @since 6.3.0
```
I don't deserve props for this, but just want to make note to update the version in all places once this is actually merged.

```suggestion
 * @since 6.4.0
```

afragen replied:
Is I'm gonna have to fix all the other instances. Not to worry 😉

dream-encode commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,499 @@
+		}
+		$body   = explode( "\n", $email['body'] );
+		$failed = __( 'These plugins failed to update or may have been restored from a temporary backup due to detection of a fatal error:' );
+		array_splice( $body, 6, 1, $failed );
```
I mentioned this to @afragen in private, but I wanted to mention it here so others could think on it.

Manually splicing the body of the email seems a bit "hacky" to me, but I don;' see a better way at the moment the way `WP_Automatic_Updater->send_plugin_theme_email` is currently structured.

The obvious drawback with the splicing method is that if/when the content and/or structure of the email changes, this line of code would also need to be updated.  If not, the addition appears in the wrong place or not at all.

It definitely not in-scope of this ticket/PR, but I am wondering if `WP_Automatic_Updater->send_plugin_theme_email` should be more hookable to enable splicing in of content at specific spots in the body text.

getsource replied:
Would it help to add a comment that further explains why / what is happening here so that it can be more easily updated in the future?

afragen replied:
It's modifying the email by adding a line. There is no other current method to conditionally modify the text of this email message.

costdev replied:
@dream-encode @getsource The `array_splice()` call has now been replaced with a `str_replace()` call in [51897c6](https://github.com/WordPress/wordpress-develop/pull/3958/commits/51897c68026f3132180d8f06a2feb3674d521875).

Both `array_splice()` and `str_replace()` aren't the most reliable of approaches.

However, the `str_replace()` call, coupled with an inline comment in `WP_Automatic_Updater::send_plugin_theme_email()` that if the string changes, it should also be changed here, should make this easy to maintain in the probably unlikely case that `__( 'These plugins failed to update:' )` is changed.

getsource replied:
That sounds good to me. Thanks!

afragen committed:
Update src/wp-admin/includes/class-rollback-auto-update.php

Co-authored-by: David B <dream-encode@users.noreply.github.com>

afragen committed:
update @ since

afragen committed:
more 6.3.0 to 6.4.0

afragen committed:
no need for Reflection here

afragen committed:
Merge branch 'trunk' into rollack-auto-update

afragen committed:
update to bring in line with feature plugin

deactivate, use include(), then reactivate. More closely follows plugin_sandbox_scrape()

afragen committed:
and not or

afragen committed:
fix email sending and bring into alignment with feature plugin

afragen committed:
add exception to defining function in main plugin file

afragen committed:
reactivate plugins at end of main loop

reactivates plugins with error exceptions

getsource commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,551 @@
+		$ref_temp_backups->setValue( $rollback_updater, $temp_backup );
+
+		// Call Rollback's restore_temp_backup().
+		$rollback_updater->restore_temp_backup();
```
I'm probably missing something here (or this is handled in the flow), but I noticed that both this and `delete_temp_backup()` return `WP_Error`.

How are these errors usually handled?
Or if it fails, is the assumption that's the best it can do?

getsource replied:
Assuming the errors are handled or expected in some way, I'd love to see a comment explaining the flow.

costdev replied:
> Or if it fails, is the assumption that's the best it can do?

That's essentially it yeah.

During an upgrade, restoration and deletion of temp backups occurs during the `'shutdown'` action, triggered in a registered shutdown function, when the upgrade _process_ fails.

While these methods _might_ fail, by this point, the previously installed version has already been backed up, the new version downloaded and unpacked to the plugins directory. We should therefore have the permissions and diskspace to do a restore and delete. PHP's `shutdown` is immune to timeouts, so that's another potential cause of failure avoided.

Nevertheless, we worded the email carefully in saying "may have been restored from a temporary backup" to account for cases that are very hard to predict given our established permissions, diskspace, and timeout immunity.

getsource replied:
Thank you! That makes sense. I would love to see this described in a comment, since it may be clear to some folks, but wasn't immediately so for me, and might make it easier for folks to work on in the future.

getsource commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,551 @@
+		$rollback_updater = new WP_Upgrader();
+
+		// Set private $temp_restores variable.
+		$ref_temp_restores = new ReflectionProperty( $rollback_updater, 'temp_restores' );
```
I noticed this seems to be the first introduction of `ReflectionProperty` in core.
Ideally we'd have an approach that avoids it, but I'm not currently sure of a better way.

Could you please add comments above the uses that make it clear why the decision was made, so that if there's a different approach that works better in the future, folks have context?

afragen replied:
It's not the first use of Reflection in core.

https://github.com/WordPress/wordpress-develop/blob/1e11c3fce87f00520d2a46322c1dc8edc9952dd7/src/wp-admin/includes/template.php#L1248-L1259

As the comment states, it's setting a private variable. I'm not sure I understand, unless you're just looking for a more verbose comment explanation. If that's the case, no problem.

The only way to avoid the Reflection is to make all the variables and methods referenced in core in the public scope.

costdev replied:
@getsource Inline comments adding context have now been added where the Reflection API is used in the class. How do they look to you?

getsource replied:
This looks great, thank you!

And thanks @afragen -- I had not found that one in my search.

afragen committed:
refactor for merge, improved naming, comments

afragen committed:
call directly without using hook

afragen committed:
cleanup

afragen committed:
oops

afragen committed:
remove array_splice

afragen committed:
¯\_(ツ)_/¯

afragen committed:
FML

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,569 @@
+	}
+
+	/**
+	 * Checks whether an error is acceptable. otherwise rolls the plugin back to its
```
```suggestion
	 * Checks whether an error is acceptable, otherwise rolls the plugin back to its
```

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,569 @@
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return array The remaining updates, excluding plugins that were processed or rolled back.
```
```suggestion
	 * @return array The remaining plugin auto-updates, excluding plugins that were processed or rolled back.
```

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,569 @@
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return array An array of remaining theme auto-updates.
```
```suggestion
	 * @return array The remaining theme auto-updates.
```

costdev committed:
Fix minor doc-block typos and inconsistencies.

costdev commented on src/wp-admin/includes/class-rollback-auto-update.php:
```diff
@@ -0,0 +1,569 @@
+/**
+ * Core class for rolling back auto-update plugin failures.
+ */
+class WP_Rollback_Auto_Update {
```
```suggestion
final class WP_Rollback_Auto_Update {
```

costdev committed:
Make `WP_Rollback_Auto_Update` `final`.

afragen committed:
trigger_error not consistent in wp-cron

mukeshpanchal27 commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1163,6 +1163,12 @@ protected function send_plugin_theme_email( $type, $successful_updates, $failed_
 			// List failed plugin updates.
 			if ( ! empty( $failed_updates['plugin'] ) ) {
+
+				/*
+				* If this string changes, the `str_replace()` call in
+				* `WP_Rollback_Auto_Update::auto_update_rollback_message()`
+				* should also be changed.
+				*/
```
```suggestion
				/*
				 * If this string changes, the `str_replace()` call in
				 * `WP_Rollback_Auto_Update::auto_update_rollback_message()`
				 * should also be changed.
				 */
```
Correct multiline comment formate.

costdev committed:
Update src/wp-admin/includes/class-wp-automatic-updater.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

afragen committed:
fix type

peterwilsoncc committed:
Merge branch 'trunk' into rollack-auto-update

afragen commented:
Closing in favor of https://github.com/WordPress/wordpress-develop/pull/5287
